src/core/memory.cc

```diff
@@ -1639,6 +1639,26 @@ void *allocate_slowpath(size_t size) {
             alloc_stats::increment(alloc_stats::types::foreign_mallocs);
             return original_malloc_func(size);
         }
+
+        // #2137 - static init fiasco for fallback functions. 
+        // If dependent libraries do malloc _before_ the above declaration inits are run,
+        // we end up here with nowhere to go. Add a second check and attempt the full init
+        // already. If we find the functions, all is good. Otherwise, trudge along and probably
+        // crash terribly later.
+        static bool double_checked = false;
+        if (!std::exchange(double_checked, true)) {
```

Comment:
This can be multi-threaded, so needs barriers.

I think we can only hit it when we are single-threaded, but at least move the write to within the if.

Reply:
Within the if? We are inside the if. 
We can only get here during dlinit phase, i.e. single threaded (assuming not loading seastar as a dynamic library, and in that case local allocator would not be permissible).
Seems pointless overkill to no either fence or atomics here.

Reply:
std::exchange() is executed outside the if, its store is unconditional.

Reply:
Why is this double_checked needed at all? If it's false, we initialize original_... stuff, if it's true _and_ original_... is null, then why not re-initialize it again?